libp2p/protocols/connectivity/autonatv2/client.nim

```diff
@@ -0,0 +1,165 @@
+      receivedNonce = dialBack.nonce, expectedNonce = expectedNonce
+    return
+
+  let resp = DialBackResponse(status: DialBackStatus.Ok)
+  try:
+    trace "Sending DialBackResponse", dialBackResp = resp
+    await conn.writeLp(resp.encode().buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    raise newException(AutonatV2Error, "Sending DialBackResponse failed", exc)
+
+proc new*(T: typedesc[AutonatV2Client], switch: Switch, rng: ref HmacDrbgContext): T =
+  let client = T(switch: switch, rng: rng, nonces: initTable[PeerId, uint64]())
+
+  # handler for DialBack messages
+  proc handleStream(
+      conn: Connection, proto: string
+  ) {.async: (raises: [CancelledError]).} =
+    try:
+      let dialBack = DialBack.decode(initProtoBuffer(await conn.readLp(1024))).valueOr:
+        raise newException(AutonatV2Error, "Unable to decode DialBack")
+      debug "Received DialBack", nonce = $dialBack.nonce
+      await client.handleDialBack(conn, dialBack)
+    except CancelledError as exc:
+      trace "handler cancelled"
+      raise exc
+    except CatchableError as exc:
+      debug "exception in handler", description = exc.msg, conn
+    except LPStreamRemoteClosedError as exc:
+      error "connection closed by peer", description = exc.msg, peer = conn.peerId
+    finally:
+      trace "closing connection", conn
+      await conn.close()
+
+  client.handler = handleStream
+  client.codec = $AutonatV2Codec.DialBack
+  client
+
+proc handleDialDataRequest*(
+    conn: Connection, req: DialDataRequest
+): Future[DialResponse] {.
+    async: (raises: [CancelledError, AutonatV2Error, LPStreamError])
+.} =
+  debug "Received DialDataRequest",
+    numBytes = req.numBytes, maxAcceptedNumBytes = MaxAcceptedDialDataRequest
+
+  if req.numBytes > MaxAcceptedDialDataRequest:
+    raise newException(
+      AutonatV2Error, "Rejecting DialDataRequest: numBytes is greater than the maximum"
+    )
+
+  # send required data
+  var msg = AutonatV2Msg(
+    msgType: MsgType.DialDataResponse,
+    dialDataResp: DialDataResponse(data: newSeq[byte](MaxDialDataResponsePayload)),
+  )
+  let messagesToSend =
+    (req.numBytes + MaxDialDataResponsePayload - 1) div MaxDialDataResponsePayload
+  for i in 0 ..< messagesToSend:
+    await conn.send(msg)
+    debug "Sending DialDataResponse", i = i, messagesToSend = messagesToSend
+
+  # get DialResponse
+  msg = AutonatV2Msg.decode(initProtoBuffer(await conn.readLp(1024))).valueOr:
+    raise newException(AutonatV2Error, "Unable to decode AutonatV2Msg")
+
+  debug "Receive message", msgType = $msg.msgType
+  if msg.msgType != MsgType.DialResponse:
+    raise
+      newException(AutonatV2Error, "Expecting DialResponse, but got " & $msg.msgType)
+
+  return msg.dialResp
+
+proc sendDialRequest*(
+    self: AutonatV2Client,
+    switch: Switch,
+    pid: PeerId,
+    peerAddrs: seq[MultiAddress],
+    testAddrs: seq[MultiAddress],
+): Future[DialResponse] {.
```

Comment:
Probably makes sense to return an object that contains the multiaddress (instead of the index), the dial status, and the reachability status similar to AutoNatV1. We also gotta be sure that we received a Dialback on the same address the server is informing us it dialed us back in the DialResponse (this should likely be done in line 154 and 156).   
```
NetworkReachability* {.pure.} = enum
    Unknown
    NotReachable
    Reachable
```

This is because we can't trust what the server is telling us. Server might say it dialed some address, but we didnt really received a dial back in such address.

Reachability should be `Reachable` on the happy path,  `NotReachable` when we receive a DIAL_ERROR, and unknown for DIAL_BACK error.

Reply:
I'll implement the check, I just think it's something to think about (maybe the spec could be improved?)